### PR TITLE
feat(streams): transactional broadcasts (PR 81)

### PR DESCRIPTION
**Pull request by mhenrixon**
Stacked on #80. **Base: \`feature/turbo-streams-phase-6\`**. First of seven follow-up PRs building out the Phase 0 roadmap features.

## Summary

**A broadcast issued inside a rolled-back \`ActiveRecord::Base.transaction\` is never delivered to any SSE client and never reaches PGMQ.** This is the feature I called out as the structurally distinctive one in the original strategic conversation — no other Rails real-time stack can offer it because every other stack (ActionCable / Redis, AnyCable, message_bus) routes broadcasts through an out-of-band pub/sub that has no knowledge of the application's transaction boundary.

\`\`\`ruby
ActiveRecord::Base.transaction do
  @order.update!(status: "shipped")
  @order.broadcast_replace_to :account           # ← deferred until commit
  RelatedService.update_counters!(@order)        # ← might raise
end
# If RelatedService raised, nothing committed AND no SSE client ever
# saw a "shipped" broadcast. Atomic with respect to each other.
\`\`\`

## Design

Pgbus detects the open transaction via \`ActiveRecord::Base.connection.current_transaction\` and calls \`transaction.after_commit { @client.send_message(@name, wrapped) }\`. AR's \`NullTransaction\` yields \`after_commit\` blocks immediately when no transaction is open, so the synchronous path (outside a transaction) degrades cleanly to the existing behavior.

**Zero new infrastructure.** No outbox table, no background worker, no extra storage. Just an \`after_commit\` block on the AR connection's own transaction object. Rollback works because the block is never called on rollback.

Full details in the commit message. The one subtle bit worth a reviewer glance: handling \`instance_double\` vs \`stub_const("ActiveRecord::Base", ...)\` — the ancestor-walk probe RSpec does for \`instance_double(Pgbus::Client)\` hits the stubbed constant and raises \`:>\` unexpected-message errors. Fix: use plain \`double\` inside the transactional describe.

## The headline integration test

\`spec/integration/streams/transactional_spec.rb\` runs three scenarios end-to-end with real AR + real PGMQ + real SSE:

1. **Commit path**: Broadcast inside \`ActiveRecord::Base.transaction\` → delivered after commit.
2. **Rollback path**: Broadcast followed by \`raise ActiveRecord::Rollback\` → SSE client sees nothing, AND \`Client#read_after\` confirms the row never landed in PGMQ.
3. **Mixed sequence**: commit → rollback → commit → exactly the committed broadcasts are delivered, in order.

All three pass in ~1 second.

## What about multi-database setups?

Users with \`connects_to: { database: { writing: :pgbus } }\` run the pgbus queue insert through a DIFFERENT AR connection than the user's main transaction. The \`after_commit\` fires on the main connection's transaction (where the user's \`@order.update!\` ran), and the callback then calls \`@client.send_message\` which owns its own pgbus connection. The broadcast lands on the right database, atomically with the user's work. Verified by unit test; integration test uses single-database setup.

## Semantics of the return value

\`Stream#broadcast\` returns the assigned \`msg_id\` when synchronous, \`nil\` when deferred. Documented trade-off: callers that need the id have to opt out of the transactional path. In practice nobody uses the return value because the whole point of the API is fire-and-forget delivery.

## Test results

- **960 unit + 9 integration = 969 passing, 0 failures**
- RuboCop clean across all touched files
- Full pgbus regression green (no cross-spec pollution after the \`instance_double\` fix)

## Test plan

- [x] \`bundle exec rspec spec/pgbus/streams_spec.rb\` — 18 examples including 8 new transactional cases
- [x] \`bundle exec rspec spec/pgbus/\` — full unit regression
- [x] \`PGBUS_DATABASE_URL=... bundle exec rspec spec/integration/streams/transactional_spec.rb\` — 3 end-to-end scenarios
- [x] \`PGBUS_DATABASE_URL=... bundle exec rspec spec/integration/streams/\` — all streams integration tests
- [x] \`bundle exec rubocop\` clean
- [ ] Reviewer: read the \`current_open_transaction\` private method — the \`rescue StandardError\` catches the "AR loaded but not connected" edge case (rake tasks before Rails boot); verify the swallow is appropriate
- [ ] Reviewer: verify the README "Transactional broadcasts" section tells the story you want to tell users

## What's next in the stack

This is PR 81 of 7 planned follow-ups:

- **#81** (this PR) — Transactional broadcasts
- #82 — Falcon streaming-body code path
- #83 — \`broadcasts_with_replay\` (chat-history-on-connect)
- #84 — Server-side audience filtering
- #85 — Presence as a stream
- #86 — Performance sweep + benchmarks
- #87 — Docs refresh

## Refs

- Original "Tier 2: Things only pgbus can do" brainstorm — the feature list that introduced this idea
- Design doc §12 roadmap item 4 — explicitly deferred in v1
- rails/rails#52420 — the bug the streams subsystem was built around; this PR is the second-most-impactful feature after the replay cursor

## Summary by CodeRabbit

* **New Features**
  * Transactional broadcasts are deferred until transaction commit and dropped on rollback; broadcasts inside transactions return nil while outside transactions remain synchronous and return a message id.

* **Documentation**
  * Added a “Transactional broadcasts” section to real-time broadcasts docs with examples and updated behavior notes (removed prior guidance about non-default transactional behavior).

* **Tests**
  * Added integration and unit tests covering commit/rollback delivery, ordering, sequential scenarios, AR/no‑AR environments, and refined test teardown behavior.

### CODE REVIEW

File: lib/pgbus/streams.rb [hunk 1]
```diff
@@ -28,9 +28,26 @@ def initialize(streamables, client: Pgbus.client)
...
+      #
+      # Transactional semantics: if this call is made inside an open
+      # ActiveRecord transaction, the PGMQ insert is deferred to an
+      # after_commit callback. If the transaction rolls back, the broadcast
+      # silently drops — clients never see the change the database never
+      # persisted. This is the feature no other Rails real-time stack
+      # (including turbo-rails over ActionCable) can offer: the broadcast
+      # and the data mutation are atomic with respect to each other.
+      # Returns the assigned msg_id when sent synchronously, nil when
+      # deferred (the id isn't known until the after_commit callback runs).
```
[the remark this answers is not in the document]
Comment (mhenrixon) [reply lib/pgbus/streams.rb:40]: Fixed in 6a0af30. Added the missing "that".

File: README.md [hunk 2]
```diff
@@ -701,9 +701,27 @@ One Puma worker hosts one `Pgbus::Web::Streamer::Instance` singleton with three
...
+end
+# If RelatedService raised, the database state is unchanged AND no SSE client
+# ever saw a "shipped" broadcast. The broadcast and the data mutation are
+# atomic with respect to each other.
+```
+
+ActionCable can't do this because its broadcast path goes through Redis pub/sub, which has no concept of your application's transaction boundary. Pgbus detects the open AR transaction via `ActiveRecord::Base.connection.current_transaction.after_commit`, which is a first-class Rails API — no outbox table, no background worker, no extra storage.
+
+Outside an open transaction, broadcasts are synchronous and return the assigned `msg_id` as before. Inside a transaction, they return `nil` (the id isn't known until commit time).
+
```
[the remark this answers is not in the document]
Comment (mhenrixon) [reply README.md:722]: Fixed in 6a0af30. Same typo fix in the README prose.

File: spec/integration/streams/transactional_spec.rb [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+    end
+
+    sleep 0.3
+    expect(client.events).to be_empty
+
+    rows = Pgbus.client.read_after(stream_name, after_id: 0, limit: 10)
+    expect(rows).to be_empty
+
+    client.close
+  end
```
[the remark this answers is not in the document]
Comment (mhenrixon) [reply spec/integration/streams/transactional_spec.rb:120]: Fixed in 6a0af30. Added a comment above the `sleep 0.3` explaining that it's a negative-assertion bound, not flaky-test ceremony — the rolled-back transaction should produce zero events and the sleep is just the wait budget for the "never happens" assertion to become reliable.

File: lib/pgbus/streams.rb [hunk 4]
```diff
@@ -74,6 +91,25 @@ def ensure_queue!
...
+
+        connection = ::ActiveRecord::Base.connection
+        transaction = connection.current_transaction
+        transaction if transaction.open?
+      rescue StandardError
+        # Defensive: if AR is loaded but not yet connected (e.g. a
+        # Rake task invoked before Rails boot), don't let the transaction
+        # probe break the broadcast.
+        nil
+      end
```
[the remark this answers is not in the document]
Comment (mhenrixon) [reply lib/pgbus/streams.rb:119]: Fixed in 4cafd73. Added a `Pgbus.logger.debug` in the rescue with the exception class and message. The CLAUDE.md coding-style rule explicitly says "NO swallowing errors — log via Pgbus.logger", so the current silent rescue was a project-rule violation. Debug is off by default in production so zero hot-path cost; when debugging "why aren't my transactional broadcasts deferring?", flipping to debug level immediately surfaces the probe failure.

File: spec/integration/streams/transactional_spec.rb [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+
+  before do
+    Pgbus.client.ensure_stream_queue(stream_name)
+    streamer.start
+  end
+
+  after do
+    streamer.shutdown!
+    harness.shutdown if defined?(@harness_started)
+  end
```
[the remark this answers is not in the document]
Comment (mhenrixon) [reply spec/integration/streams/transactional_spec.rb:72]: Fixed in 4cafd73. Also normalised the same pattern in `reconnect_spec.rb` and `page_born_stale_spec.rb` so the three streams integration specs stay consistent. `connect_sse_client` only ever sets `@harness_started = true`, never to `nil`/`false`, so the `defined?` guard was doing no work beyond what a plain truthy check does.

File: spec/integration/streams/transactional_spec.rb [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+    Pgbus.configuration.streams_heartbeat_interval = 30
+    Pgbus.configuration.streams_write_deadline_ms = 5_000
+    Pgbus.reset_client!
+  end
+
+  after(:all) do
+    Pgbus.configuration.listen_notify = @saved_listen_notify
+    Pgbus.configuration.streams_signed_name_secret = nil
+    Pgbus.reset_client!
+  end
```
[the remark this answers is not in the document]
Comment (mhenrixon) [reply spec/integration/streams/transactional_spec.rb:41]: Fixed in e3181d2. Normalised both `transactional_spec.rb` and `page_born_stale_spec.rb` to match the reconnect_spec pattern (save all five mutated stream config values in `before(:all)`, restore all five in `after(:all)`). This is the same fix PR #77 review applied to reconnect_spec; I missed the other two specs at the time because phase-7 didn't touch them. All three streams integration specs now use the same save-all/restore-all template.
